commands/ls.go

```diff
@@ -68,3 +81,49 @@ func cmdLs(c *cli.Context) {
 
 	w.Flush()
 }
+
+func parseFilters(filters []string) (FilterOptions, error) {
+	options := FilterOptions{}
+	for _, f := range filters {
+		kv := strings.SplitN(f, "=", 2)
+		key, value := kv[0], kv[1]
+
+		switch key {
+		case "swarm":
+			options.SwarmName = value
+		case "driver":
+			options.DriverName = value
+		default:
+			return options, fmt.Errorf("Unsupported filter key '%s'", key)
+		}
+	}
+	return options, nil
+}
+
+func filterHosts(hosts []*libmachine.Host, filters FilterOptions) []*libmachine.Host {
+	filteredHosts := []*libmachine.Host{}
+	swarmMasters := make(map[string]string)
+	for _, h := range hosts {
+		swarmOptions := h.HostOptions.SwarmOptions
+		if swarmOptions != nil && swarmOptions.Master {
+			swarmMasters[swarmOptions.Discovery] = h.Name
+		}
+	}
+
+	for _, h := range hosts {
```

Comment:
To clarify, I generally try to avoid deeply nested blocks of logic, and also I have a hard time reasoning about what this little chunk does just by looking at it, so it'd be good to have it in a little function called `getX` and I can look at the assignment and say, "Oh, this is the bit that gets X from these pieces of information passed in".